File: src/main/java/github/macro/build_info/BuildSerializer.java
package github.macro.build_info;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;
import github.macro.build_info.equipment.EquipmentInfo;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;

/**
 * Created by Macro303 on 2020-Jan-23.
 */
public class BuildSerializer extends StdSerializer<Build> {
	private static final Logger LOGGER = LogManager.getLogger(BuildSerializer.class);

	public BuildSerializer() {
		this(null);
	}

	public BuildSerializer(Class<Build> t) {
		super(t);
	}

	@Override
	public void serialize(Build value, JsonGenerator parser, SerializerProvider provider) throws IOException {
		parser.writeStartObject();
		parser.writeStringField("Version", value.getVersion());
		parser.writeStringField("Name", value.getName());
		parser.writeStringField("Class", value.getClassTag().name());
		parser.writeStringField("Ascendency", value.getAscendency().name());
		parser.writeObjectField("Gems", value.getBuildGems());
		parser.writeObjectField("Equipment", value.getEquipment().stream().map(EquipmentInfo::getName));
		parser.writeEndObject();
	}
}

File: src/main/kotlin/github/macro/build_info/BuildSerializer.kt
package github.macro.build_info

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import org.apache.logging.log4j.LogManager
import java.io.IOException

/**
 * Created by Macro303 on 2020-Jan-13.
 */
class BuildSerializer @JvmOverloads constructor(t: Class<Build>? = null) : StdSerializer<Build>(t) {

    @Throws(IOException::class, JsonProcessingException::class)
    override fun serialize(value: Build, parser: JsonGenerator, provider: SerializerProvider?) {
        parser.writeStartObject()
        parser.writeStringField("name", value.name)
        parser.writeStringField("class", value.classTag.name)
        parser.writeStringField("ascendency", value.ascendency.name)
        parser.writeObjectFieldStart("gems")
        value.gemBuild.let {
            parser.writeObjectField("links", it.links.map { link -> link.map { it?.getFullname() } })
            parser.writeArrayFieldStart("updates")
            it.updates.forEach { update ->
                parser.writeStartObject()
                parser.writeStringField("oldGem", update.oldGem?.getFullname())
                parser.writeStringField("newGem", update.newGem?.getFullname())
                parser.writeStringField("reason", update.reason)
                parser.writeEndObject()
            }
            parser.writeEndArray()
        }
        parser.writeEndObject()
        parser.writeObjectField("equipment", value.equipment.map { it.name })
        parser.writeEndObject()
    }

    companion object {
        private val LOGGER = LogManager.getLogger(BuildSerializer::class.java)
    }
}